test/test_utils/GenericRequestResponseHandler.h

```diff
@@ -0,0 +1,128 @@
+// Copyright 2004-present Facebook. All Rights Reserved.
+
+#pragma once
+
+// #include "RSocketTests.h"
+#include "yarpl/Single.h"
+
+#include "folly/ExceptionWrapper.h"
+
+using namespace yarpl;
+using namespace yarpl::single;
+using namespace rsocket;
+using namespace rsocket::tests;
+using namespace rsocket::tests::client_server;
+
+namespace rsocket { namespace tests {
+
+using StringPair = std::pair<std::string, std::string>;
+
+std::ostream& operator<<(std::ostream& os, StringPair const& payload) {
+  return os << "('" << payload.first << "', '" << payload.second << "')";
+}
+
+struct ResponseImpl {
+  using Response = std::shared_ptr<ResponseImpl>;
+  using MapFunc = std::function<Response(StringPair const&)>;
+
+  enum Type {
+    P, E, F
+  };
+
+  StringPair p;
+  std::exception_ptr e;
+  MapFunc f;
+  Type type;
+
+  std::function<void(StringPair const&)> before_handle;
+
+  explicit ResponseImpl(StringPair const& p) : p(p), type(Type::P) {}
+  explicit ResponseImpl(std::exception_ptr const& e) : e(e), type(Type::E) {}
+  explicit ResponseImpl(MapFunc lam) : f(lam), type(Type::F) {}
+
+  ~ResponseImpl() {}
+};
+
+using Response = ResponseImpl::Response;
+using MapFunc  = ResponseImpl::MapFunc;
+
+struct GenericRequestResponseHandler : public rsocket::RSocketResponder {
+  using ResponseMap = std::unordered_map<StringPair, Response>;
+  // Response can be a payload, or an exception (which is rethrown)
+
+  GenericRequestResponseHandler(ResponseMap&& responses, Response def) : 
+    responses_(std::move(responses)), default_(def) {}
+
+  GenericRequestResponseHandler(Response def) :
+    default_(def) {}
+
+  Reference<Single<Payload>> handleRequestResponse(Payload request, StreamId) 
+  override 
+  {
+    auto data = request.moveDataToString();
+    auto meta = request.moveMetadataToString();
+
+    Response resp;
+    StringPair req = StringPair(data, meta);
+    auto in_map = this->responses_.find(req);
+    if (in_map == this->responses_.end()) {
+      resp = this->default_;
+    }
+    else {
+      resp = in_map->second;
+    }
+
+    // recursively resolve until response isn't a function
+    while(resp->type == ResponseImpl::Type::F) {
+      resp = resp->f(req);
+    }
+
+    return Single<Payload>::create(
+      [resp, this](auto subscriber)
```

Comment:
Nit: Are you using `clang-format`?  If not you should be, this styling isn't consistent with the rest of the codebase.